### PR TITLE
Add React hooks for connecting to the API

### PR DESCRIPTION
**Pull request by controversial**
This PR adds `useApiData` and `useApiRequest` hooks for easily requesting data from the API.

Note: this depends on #88; that PR should be merged before this one.

## Usage
To fetch data to use in a component, simply use:
```tsx
import { useApiData } from '../../api';

const MyComponent: React.FC = () => {
  const { data, error } = useApiData('api-route'); // 'data' and 'error' are either Objects or `undefined`; never promises or anything like that
  return (
    <div>
      { JSON.Stringify(data) }
    </div>
  );
}
```
The `useApiData` hook uses [`swr`](https://github.com/zeit/swr) to implement caching, revalidation, and automatic retries.

To manually make API requests in a component without any of these features, use the `useApiRequest` hook instead.
```tsx
import { useApiRequest } from '../../api';

const MyComponent: React.FC = () => {
  const sendReq = useApiRequest('api-route', { method: 'POST', body: 'hello' });
  return (
    <button type="button" onClick={sendReq}>
      click me
    </button>
  );
}
```

## Details 

Under the hood, it uses Zeit's [`swr`](https://github.com/zeit/swr) library along with a [`fetch` ponyfill](https://github.com/developit/unfetch). 

Right now, there is *no* support for authentication, so all requests return the same JSON response: 
```json
{"detail":"Authentication credentials were not provided."}
```

A future (very soon) PR, which implements the client authentication flow, will update this `useApi` hook to automatically leverage API tokens added to the global application store.

### CODE REVIEW

File: client/src/api.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import useSWR, { responseInterface } from 'swr';
+import fetch from 'unfetch';
```
Comment (duci9y) [inline client/src/api.ts:2]: No need since latest Chrome is our target!

File: client/src/api.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  body?: FormData | Blob | ArrayBuffer | string;
+  addTrailingSlash?: boolean;
+}
+
+/**
+ * A react hook for connecting a component to API data
+ * Automatically implements caching, revalidation, retries, and some other stuff. To make API
+ * requests without any of these features, see useApiRequest
+ */
+export function useApiData<T>(
```
Comment (duci9y) [inline client/src/api.ts:18]: Maybe we should force `const name = () => {}` style in eslint for consistency?

File: client/src/api.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import useSWR, { responseInterface } from 'swr';
+import fetch from 'unfetch';
+
+import { useStore } from './store';
+
+interface apiRequestOptions {
```
Comment (duci9y) [inline client/src/api.ts:6]:
```suggestion
interface ApiRequestOptions {
```

File: client/src/pages/Home/index.tsx [hunk 4]
```diff
@@ -1,9 +1,13 @@
 import React from 'react';
+import { useApiData } from '../../api';
```
Comment (duci9y) [inline client/src/pages/Home/index.tsx:2]: Use absolute imports.

File: client/src/api.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ * Make a request to the API
+ * @param base - the base URL (for example: https://localhost:8000/api/v1/)
+ * @param path - the path after that URL (for example: 'token')
+ * @param options - options to apply to the fetch request
+ */
+export function apiReq(base: string, path: string, options: apiRequestOptions) {
+  // Add the path to the base URL
+  let { href } = new URL(path, base);
+  // Unless addTrailingSlash was explicitly passed as false, we make sure it's there
+  if (options.addTrailingSlash !== false && !href.endsWith('/'))
```
Comment (duci9y) [inline client/src/api.ts:61]: It looks like this file didn't go through prettier?

File: client/src/api.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    href = `${href}/`;
+
+  return fetch(href, options).then(async (r) => {
```
Comment (duci9y) [inline client/src/api.ts:64]: Use Promise chaining.

File: client/src/api.ts (hunk 1 above)
Comment (controversial) [reply client/src/api.ts:2]: got it, thanks! I'll take it out

File: client/src/api.ts (hunk 6 above)
Comment (controversial) [reply client/src/api.ts:64]: Do you mean instead of the async function?

File: client/src/api.ts (hunk 5 above)
Comment (controversial) [reply client/src/api.ts:61]: I think it did, and it also passes linting; what looks wrong with it?

**Pull request merged**